proposals/subgroups.md

```diff
@@ -0,0 +1,227 @@
+this feature set captures ~84% of devices.
+Splitting the feature does not grab a significant portion of devices.
+Splitting out simd-scoped reduction adds the Apple6 family which includes
+iPhone11, iPhone SE, and iPad (9th gen).
+
+**TODO**: Should there be additional enables for extra functionality?
+Some possibilities:
+* MSL and HLSL (SM6.8) support `any` and `all` operations at quad scope
+* SPIR-V and HLSL (SM6.5) could support more exclusive/inclusive, clustered, and partitioned operations
+* Inclusive sum and product could be done with multi-prefix SM6.5 operations in HLSL or through emulation
+
+## Built-in Values
+
+| Built-in | Type | Direction | Description |
+| --- | --- | --- | --- |
+| `subgroup_size` | u32 | Input | The size of the current subgroup |
+| `subgroup_invocation_id` | u32 | Input | The index of the invocation in the current subgroup |
+
+Note: HLSL does not expose a subgroup_id or num_subgroups equivalent.
+
+## Built-in Functions
+
+All built-in function can only be used in `compute` or `fragment` shader stages.
+Using f16 as a parameter in any of these functions requires `subgroups-f16` to be enabled.
+
+| Function | Preconditions | Description |
+| --- | --- | --- |
+| `fn subgroupElect() -> bool` | | Returns true if this invocation has the lowest subgroup_invocation_id among active invocations in the subgroup |
+| `fn subgroupAll(e : bool) -> bool` | | Returns true if `e` is true for all active invocations in the subgroup |
+| `fn subgroupAny(e : bool) -> bool` | | Returns true if `e` is true for any active invocation in the subgroup |
+| `fn subgroupBroadcast(e : T, id : I) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types<br>`I` must be i32 or u32 | Broadcasts `e` from subgroup_invocation_id `id` to all active invocations. `id` must be dynamically uniform<sup>1</sup> |
+| `fn subgroupBroadcastFirst(e : T) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types | Broadcasts `e` from the active invocation with the lowest subgroup_invocation_id in the subgroup  to all other active invocations |
+| `fn subgroupBallot(pred : bool) -> vec4<u32>` | | Returns a set of bitfields where the bit corresponding to subgroup_invocation_id is 1 if `pred` is true for that active invocation and 0 otherwise. |
+| `fn subgroupShuffle(v : T, id : I) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types<br>`I` must be u32 or i32 | Returns `v` from the active invocation whose subgroup_invocation_id matches `id` |
+| `fn subgroupShuffleXor(v : T, mask : u32) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types | Returns `v` from the active invocation whose subgroup_invocation_id matches `subgroup_invocation_id ^ mask` |
+| `fn subgroupShuffleUp(v : T, delta : u32) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types | Returns `v` from the active invocation whose subgroup_invocation_id matches `subgroup_invocation_id - delta` |
```

Comment:
What happens when subgroup_invocation_id +- delta goes OOB?

Reply:
An indeterminate value is returned. I was trying to avoid all the fiddly cases in the proposal that will have to be addressed in the final spec.

MSL says lower invocations are unmodified, but SPIR-V says undefined value.

Reply:
Actually MSL has a variant of shuffles which fill-in a defined value in the OOB cases. But leaving it as UB is fine as well IMO.

Reply:
Maybe I'm misunderstanding this, but given that at least one invocation would end up with an indeterminate value, wouldn't the developer always have to end up doing some comparison against `subgroup_invocation_id` to ensure the indeterminate value doesn't cause trouble?